### PR TITLE
Add (most of) the ODRI Coding Guidelines

### PR DESCRIPTION
**Pull request by luator**
## Description

Integrate the [ODRI codeing guidelines](https://github.com/machines-in-motion/package_template/tree/master/doc)
and merge it with the guidelines we already had here.

Concrete formatting rules are omitted as the rule is simply "use the
auto-formatter".

Also modified a bit where I found it appropriate (e.g. suggested using
flake8 when working with Python) and integrated the things discussed on
our last meeting (which repos do need code review).

Let me know what you think.

## How I tested
Built the documentation locally.

### CODE REVIEW

File: doc/E2_guidelines.md [hunk 1]
```diff
@@ -1,96 +1,344 @@
...
+
+### Creating a pull request
+
+- Visit the repository on GitHub,
+- click on the link "Pull requests",
+- create a new pull request, assign yourself and select reviewers using the web
+  interface,
 - select an appropriate label.
 
 It is good manner to squash all the commits of the branch before a code review (see below), as this makes the work of the reviewer easier. See for example these [instructions](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html).
```
Comment (luator) [inline doc/E2_guidelines.md]: How does this actually make reviewing easier?  I would rather suggest that sometimes it can help to have separate commits that can be reviewed one by one (however, in this case, it should probably rather be separate PRs).

Comment (vincentberenz) [reply doc/E2_guidelines.md]: Ideally one push request should be about the change/improvement of only one feature, so having only one commit would make sense. 
On my side I typically push on the branch on my desktop in order to keep working on my laptop (or the opposite), and having related commit is just noise for the reviewer.
But maybe that should be more of a case by case issue that a general rule.
We can rephrase:

"When relevant, it is good manner ..."

**Review by vincentberenz — APPROVED** [review]
Thanks a lot for the update.
Once merged, I'll generate the html and upload the files.
Currently I do this via colcon/mpi-cmake-modules. 
But since this package does not have any dependencies nor any code, maybe we could use github to generate the documentation ? If you agree, I'll look into this.

**Comment by vincentberenz** [comment]
> Thanks a lot for the update. Once merged, I'll generate the html and upload the files. Currently I do this via colcon/mpi-cmake-modules. But since this package does not have any dependencies nor any code, maybe we could use github to generate the documentation ? If you agree, I'll look into this.

Only saw now that you opened an issue saying exactly this :)

**Review by vincentberenz — APPROVED** [review]
That does motivate me a lot.

File: doc/E2_guidelines.md (hunk 1 above)
Comment (luator) [reply doc/E2_guidelines.md]: I see your point.  How about this?

```suggestion
It is good manner to squash commits that logically belong together (e.g. multiple small changes on the same part of code) before a code review (see below), as this makes the work of the reviewer easier and leads to a cleaner git history. See for example these [instructions](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html).
```

Comment (vincentberenz) [reply doc/E2_guidelines.md]: +1

**Comment by luator** [comment]
@dtrb @simonguist I merged now as this was blocking another PR, please still let me know in case you have comments.
